Sources/Hackle/Core/Database/Database.swift

```diff
@@ -4,78 +4,91 @@ class Database {
     private let lock: ReadWriteLock
     private let filepath: String?
     private(set) var version: Int
+    private let versionRepository = UserDefaultsKeyValueRepository.of(suiteName: storageSuiteNameVersion)
 
     init(label: String, filename: String, version: Int) {
         self.lock = ReadWriteLock(label: label)
         let manager = FileManager.default
         let url = manager.urls(for: .libraryDirectory, in: .userDomainMask).last
-        filepath = url?.appendingPathComponent(filename).path
+        self.filepath = url?.appendingPathComponent(filename).path
         self.version = version
 
-        let currentVersion = self.getVersion()
-        // 현재 버전이 0이면 DB 생성이 안되어 있거나, 버전 명시가 안된 구버전 DB 이다
-        // 구버전 DB의 마이그레이션은 onCreate에서 직접 진행해야 한다.
-        if currentVersion == Database.DEFAULT_VERSION {
-            onCreate()
-        } else if currentVersion < self.version {
-            onUpdate(oldVersion: currentVersion, newVersion: self.version)
+        do {
+            let currentVersion = getVersion(key: filename)
+            let ddls = getDDLs(oldVersion: currentVersion, newVersion: version)
+            try executeDDLs(ddls: ddls)
+        } catch {
+            do {
+                Log.error("drop and recreate database \(label), because of error: \(error)")
+                try onDrop()
+                let ddls = getDDLs(oldVersion: Database.DEFAULT_VERSION, newVersion: version)
+                try executeDDLs(ddls: ddls)
+            } catch {
+                Log.error("failed to create database \(label), because of error: \(error)")
+                setVersion(key: filename, version: Database.DEFAULT_VERSION)
+                return
+            }
         }
-        
-        // 테이블 생성/업데이트 완료 후 버전을 DB에 반영
-        setVersion()
+
+        setVersion(key: filename, version: self.version)
     }
     
+    /// 쿼리를 실행합니다.
+    /// - Parameter command: 쿼리
+    /// - Returns: 리턴값
     func execute<T>(command: (SQLiteDatabase) throws -> T) rethrows -> T {
         try lock.write {
             try SQLiteDatabase(databasePath: filepath).use { database in
                 try command(database)
             }
         }
     }
-    
-    /// table을 생성해야할 때 호출됩니다.
-    ///
-    /// 이 메소드를 override하여 table을 생성합니다.
-    func onCreate() {
-        // nothing to do
+
+    /// oldVersion ~ newVersion 사이의 DDL을 반환합니다.
+    /// - Parameters:
+    ///  - oldVersion: 이전 버전
+    ///  - newVersion: 새 버전
+    ///  - Returns: DDL 리스트
+    func getDDLs(oldVersion: Int, newVersion: Int) -> [DatabaseDDL] {
```

Comment:
```suggestion
    func getDDLs() -> [DatabaseDDL] {
```
oldVersion, newVersion 체크를 각 구현체마다 해야 하는데 Database에서 공통 관리하는게 좋을듯합니다